art/attacks/poisoning/feature_collision_attack.py

```diff
@@ -0,0 +1,279 @@
+# WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+# TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+"""
+This module implements clean-label attacks on Neural Networks.
+"""
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+import logging
+
+import numpy as np
+from tqdm import tqdm
+
+from art.attacks.attack import PoisoningAttackWhiteBox
+from art.estimators import BaseEstimator, NeuralNetworkMixin
+
+from functools import reduce
+
+logger = logging.getLogger(__name__)
+
+
+class FeatureCollisionAttack(PoisoningAttackWhiteBox):
+    """
+    Close implementation of Feature Collision Poisoning Attack by Shafahi, Huang, et al 2018.
+    "Poison Frogs! Targeted Clean-Label Poisoning Attacks on Neural Networks"
+
+    | Paper link: https://arxiv.org/pdf/1804.00792.pdf
+    """
+
+    attack_params = PoisoningAttackWhiteBox.attack_params + [
+        "classifier",
+        "target",
+        "feature_layer",
+        "learning_rate",
+        "decay_coeff",
+        "stopping_tol",
+        "num_old_obj",
+        "max_iter",
+        "similarity_coeff",
+        "watermarking",
+    ]
+
+    _estimator_requirements = (BaseEstimator, NeuralNetworkMixin)
+
+    def __init__(self, classifier, target, feature_layer, learning_rate=500 * 255.0, decay_coeff=0.5,
+                 stopping_tol=1e-10, num_old_obj=40, max_iter=120, similarity_coeff=0.25, watermark=0.3,
+                 **kwargs):
+        """
+        Initialize an SVM poisoning attack
+
+        :param classifier: A trained neural network classifier
+        :type classifier: (`art.estimators.NeuralNetworkMixin`, `art.estimators.BaseEstimator`)
+        :param target: The target input to misclassify at test time
+        :type target: `np.ndarray`
+        :param feature_layer: The name of the feature representation layer
+        :type feature_layer: `str`
+        :param learning_rate: The learning rate of clean-label attack optimization
+        :type learning_rate: `float`
+        :param decay_coeff: The decay coefficient of the learning rate
+        :type decay_coeff: `float`
+        :param stopping_tol: The tolerance for relative change in objective function
+        :type stopping_tol: `float`
+        :param num_old_obj: The number of old objective values to store
+        :type num_old_obj: `int`
+        :param max_iter: The maximum number of iterations for the attack
+        :type max_iter: `int`
+        :param similarity_coeff: The maximum number of iterations for the attack
+        :type similarity_coeff: `float`
+        :param watermark: Whether The opacity of the watermarked target image
+        :type watermark: `float`
+        :param kwargs: Extra optional keyword arguments
```

Comment:
What is using these kwargs? Are they required?

Reply:
`kwargs` is used by `set_params`. Here I'm just following conventions from other attacks. `set_params` actually validates that variables match certain constraints.

Reply:
I think this is where I got confused, this is different from evasion and the other attacks. The idea of line 43 is to make sure `set_params` can only add expected attributes, which are all mention explicitly in `__init__`. Would it be possible to set all attributes following the approach with the evasion attacks?
I know the approach in lines 90-98 facilitates type checking, etc., but we are updating `set_params` in #360 to fix this.